### ISSUE TITLE
Self-hosted runners for CI

### ISSUE THREAD

**Issue by poodlewars** [man-group/ArcticDB#941]
Set up self-hosted runners for our CI workflows, rather than using the free ones that Github provides.

We found the Github provided "large" runners prohibitively expensive (hundreds of thousands of dollars per year).

So we can look at:

- AWS or Azure hosted runners (either is fine, this is the first preference as long as cost not in to hundreds of thousands per year)
- On-prem hosted runners within Man Group (Linux quite easy, could reuse our "build servers"? Windows harder as those build agents are more locked down). Could even consider just buying some of our own boxes and running them ourselves rather than in a datacentre. Only do this if the cloud hosted versions are prohibitively expensive.

We want to do the above for Linux and Windows.

For Mac, we have some Mac Studio boxes in the post and we'll use those.

Relates to #481

**Comment by G-D-Petrov** [comment]
After looking into this over the past week, I have made some findings that I would like to document here.
Given how our build and tests are currently set up, using self hosted runners wouldn't help much.
This is because there are 2 major limitations currently:
1. The build having too many dependencies and templates which results in big binaries which overrun the cache
2. The performance of the tests on Windows, especially the integration tests (e.g on [this build](https://github.com/man-group/ArcticDB/actions/runs/6661699067/usage) the integration tests for python 3.8 took ~9 min on linux and **~69 min on windows**)
- The performance discrepancy is similar when testing on my headnotes, so this suggest a general problem with Windows I/O and getting better runners is unlikely to help much

For the first issue, here are some comparisons of random builds:
Build| Fetch dependencies | Build | Total
-- | -- | -- | --
[master(with cheap runners)](https://github.com/man-group/ArcticDB/actions/runs/6661699067/job/18104956613)| 159s | 174s | 333s 
[build with ec2 runner](https://github.com/man-group/ArcticDB/actions/runs/6664766153/job/18113068246)|  214s |  96s | 310s 
[build with GHA Large runner (with cold cache)](https://github.com/man-group/ArcticDB/actions/runs/6666764980/job/18118937676)|  63s |  490s | 553s 
[build with a GHA Large runner (with hot cache)](https://github.com/man-group/ArcticDB/actions/runs/6651991136/job/18075148891)|  362s |  49s | 411s

As you can see, there is a lot of variability and we are at the mercy of the networking gods .
Of course, over the long run, the better machines will perform better but for any given build, it is a toss-up if the hardware will actually help.

For the second issue, I gave the example of the integration tests because currently it is the major bottleneck.
There is also the issue with the hypothesis tests which again take a long time.
And in some cases, also the unit tests, because there are hypothesis tests mixed in there as well.

Based on all of the above, I think that there are a few things that we can do in order to speed-up the builds:
1. Look into increasing the sccache limit of 10GB - if this can be resolved, it would have by far the most impact
2. Set-up an ec2 runner to build and test the code as quickly as possible, in order to give some feedback to the develop quickly - [already done in this PR](https://github.com/man-group/ArcticDB/pull/997)
3. Look into using better runnners for the compilations(this means also the wheels!) and regular ones for the tests
- this way we get the best of both worlds - compilations that are as fast as possible and we don't spend too much money on the tests
- [I already experimented with this here](https://github.com/man-group/ArcticDB/actions/runs/6666764980/usage) only with the Linux build and it is performing about the same as using only large runners but at only ~40% of the cost 
4. Reduce the amount of integration tests on windows and test only against in-memory storage
-[ I already experimented with something similar here ](https://github.com/man-group/ArcticDB/actions/runs/6664766153)and it helped reduce the test time a lot but here there were still tests against Azure
5. Reduce the dependencies and the use of templates - I don't think that we have any plans to do this as the moment


TL;DR - if we do steps 2-4, we can reduce the total build time to around 40-50 min (currently 100-120min)
If we also fix the sccache issue (step 1), we could take it down even further to ~20-25min when the cache is hot.

**Comment by poodlewars** [comment]
For 1, would it help to use different vcpkg [binary caching](https://learn.microsoft.com/en-us/vcpkg/users/binarycaching)? For example we could set up an S3 bucket (or whatever else is supported) as the binary cache for VCPKG. We could do the [same ](https://github.com/mozilla/sccache/blob/main/docs/S3.md) with `sccache` for our own artifacts too. Then we would effectively have an unlimited cache size (we would need to bound it to some large value, but definitely more than 10GB).

2 and 3 both sound good.

For 4, reducing build times on Windows sounds like something best done with code / test changes. Something we can consider after our suppose. At least getting fast Linux builds would be a good start.
